ecllibrary/std/Date.ecl

```diff
@@ -977,6 +977,70 @@ EXPORT Date_t AdjustCalendar(Date_t date,
     TimeLib.AdjustCalendar(date, year_delta, month_delta, day_delta);
 
 
+/**
+ * Helper function.  Calculates the 1-based week number of a date, starting from
+ * a reference date.
+ *
+ * @param date              The date for which to compute the week number;
+ *                          must be greater than or equal to referenceDate
+ * @param referenceDate     The date from which the week number counting begins;
+ *                          must be less than or equal to date
+ * @param startingDayOfWeek The index number of the first day of a week, 1-7,
+ *                          where 1 = Sunday
+ * @return                  The 1-based week number of date, relative to
+ *                          referenceDate
+ *
+ * @see YearWeekNumFromDate, MonthWeekNumFromDate
+ */
+
+SHARED WeekNumForDate(Date_t date, Date_t referenceDate, UNSIGNED1 startingDayOfWeek) := FUNCTION
+    referenceDayOfWeek := DayOfWeek(referenceDate);
+    startingDayOfWeekDelta := (startingDayOfWeek - referenceDayOfWeek) % 7;
+    referenceFirstDateOfWeek := AdjustDate(referenceDate, day_delta := startingDayOfWeekDelta);
+    numberOfDays := DaysBetween(referenceFirstDateOfWeek, date) + 1;
+    weekNum0 := (numberOfDays + 6) DIV 7;
+    weekNum := IF(startingDayOfWeek > referenceDayOfWeek, weekNum0 + 1, weekNum0);
+
+    RETURN weekNum;
+END;
+
+/**
+ * Returns the 1-based week number of a date within the date's year.
+ *
+ * @param date              The date for which to compute the week number
+ * @param startingDayOfWeek The index number of the first day of a week, 1-7,
+ *                          where 1 = Sunday; OPTIONAL, defaults to 1
+ * @return                  The 1-based week number of date, relative to
+ *                          the beginning of the date's year
+ *
+ * @see MonthWeekNumFromDate
+ */
+
+EXPORT YearWeekNumFromDate(Date_t date, UNSIGNED1 startingDayOfWeek = 1) := FUNCTION
```

Comment:
startingDayOfWeek is optional in the exported functions, defaulting to Sunday.  The internal (shared) function has no optional arguments because all callers should supply all of them.

Reply:
Ok